src/Raven.Studio/typescript/widgets/virtualGrid/VirtualGrid.ts

```diff
@@ -0,0 +1,382 @@
+            for (var i = firstVisibleRowIndex; i < firstVisibleRowIndex + totalVisible; i++) {
+                if (!this.items[i]) {
+                    needsToFetch = true;
+                    break;
+                }
+            }
+        }
+
+        // We're missing items. Fetch them.
+        if (needsToFetch) {
+            // Take about 50 items before the first visible index (in case we're scrolling up)
+            var skip = Math.max(0, firstVisibleRowIndex - 50);
+            var take = Math.max(VirtualGrid.minItemFetchCount, totalVisible);
+            this.fetchItems(skip, take);
+        }
+    }
+
+    private findRowAtY(y: number): VirtualRow | null {
+        // This is hot path, called multiple times when scrolling. 
+        // Keep allocations to a minimum.
+        for (var i = 0; i < this.virtualRows.length; i++) {
+            var vRow = this.virtualRows[i];
+            var vRowTop = vRow.top;
+            var vRowBottom = vRowTop + VirtualRow.height;
+            if (vRowTop <= y && vRowBottom > y) {
+                return vRow;
+            }
+        }
+
+        return null;
+    }
+
+    private getOffscreenRow(viewportTop: number, viewportBottom: number): VirtualRow {
+        // This is hot path, called multiple times when scrolling.
+        // Keep allocations to a minimum.
+        for (var i = 0; i < this.virtualRows.length; i++) {
+            var row = this.virtualRows[i];
+            if (row.isOffscreen(viewportTop, viewportBottom)) {
+                return row;
+            }
+        }
+
+        throw new Error(`Bug: couldn't find an offscreen row to recycle. viewportTop = ${viewportTop}, viewportBottom = ${viewportBottom}, recycle row count = ${this.virtualRows.length}`);
+    }
+
+    private chunkFetchFailed(error: any, skip: number, take: number) {
+        // Any rows displaying these items, show an error.
+        var endIndex = skip + take;
+        var failedRows = this.virtualRows
+            .filter(r => !r.hasData && r.index >= skip && r.index <= endIndex);
+        failedRows.forEach(r => r.dataLoadError(error));
+    }
+
+    private chunkFetched(results: PagedResult) {
+        if (!this.columns() || this.columns().length === 0) {
+            this.assignColumnFromItems(results.items);
+        }
+
+        // Add these results to the .items array as necessary.
+        this.items.length = results.totalCount;
+        this.totalItemCount = results.totalCount;
+        this.virtualHeight(results.totalCount * VirtualRow.height);
+        var endIndex = results.skip + results.items.length;
+        for (var i = 0; i < results.items.length; i++) {
+            var rowIndex = i + results.skip;
+            this.items[rowIndex] = results.items[i];
+        }
+
+        this.redraw();
+    }
+
+    private assignColumnFromItems(items: any[]) {
+        var propertySet = {};
+        items.forEach(i => Object.keys(i).forEach(prop => (<any>propertySet)[prop] = true));
```

Comment:
If it is only internal usage you can use es6 set or map (we already have shim for this). Unfortunately we can't use es6 set/maps with knockout.